agreement_legal_sale/models/sale_order.py

```diff
@@ -17,34 +17,50 @@ def _action_confirm(self):
         for order in self:
             if order.agreement_template_id:
-                order.agreement_id = order.agreement_template_id.copy(default={
-                    'name': order.name,
-                    'code': order.name,
-                    'is_template': False,
-                    'sale_id': order.id,
-                    'partner_id': order.partner_id.id,
-                    'analytic_account_id': order.analytic_account_id and
-                    order.analytic_account_id.id or False,
-                })
+                order.agreement_id = order.\
+                    agreement_template_id.copy(default={
+                        'name': order.name,
+                        'code': order.name,
+                        'is_template': False,
+                        'sale_id': order.id,
+                        'partner_id': order.partner_id.id,
+                        'analytic_account_id': order.analytic_account_id and
+                        order.analytic_account_id.id or False,
+                    })
                 for line in order.order_line:
                     # Create agreement line
-                    self.env['agreement.line'].create({
-                        'product_id': line.product_id.id,
-                        'name': line.name,
-                        'agreement_id': order.agreement_id.id,
-                        'qty': line.product_uom_qty,
-                        'sale_line_id': line.id,
-                        'uom_id': line.product_uom.id
-                    })
-                    # If the product creates service profiles, create one
-                    if line.product_id.product_tmpl_id.is_serviceprofile:
-                        self.env['agreement.serviceprofile'].create({
-                            'name': line.name,
-                            'product_id': line.product_id.product_tmpl_id.id,
-                            'agreement_id': order.agreement_id.id,
-                        })
+                    self.env['agreement.line'].\
+                        create(self._get_agreement_line_vals(line, order))
+                    # Create SP's based on product_id config
+                    if line.product_id.is_serviceprofile:
+                        self.create_sp_qty(line, order)
         return res
 
+    def create_sp_qty(self, line, order):
+        # Create line.product_uom_qty SP's
+        if line.product_id.product_tmpl_id.is_serviceprofile:
+            for i in range(1, int(line.product_uom_qty)+1):
+                sp = self.env['agreement.serviceprofile'].\
+                    create(self._get_sp_vals(line, order))
+                sp.write({'name': sp.name + ' ' + str(i)})
+
+    def _get_agreement_line_vals(self, line, order):
+        return {
+            'product_id': line.product_id.id,
+            'name': line.name,
+            'agreement_id': order.agreement_id.id,
+            'qty': line.product_uom_qty,
```

Comment:
The `order` argument is not necessary:
```suggestion
    def _get_agreement_line_vals(self, line):
        return {
            'product_id': line.product_id.id,
            'name': line.name,
            'agreement_id': line.order_id.agreement_id.id,
            'qty': line.product_uom_qty,
```